org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/ServiceUseRule.java

```diff
@@ -0,0 +1,212 @@
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.osgi.framework.Filter;
+import org.osgi.framework.ServiceReference;
+import org.osgi.test.common.osgi.ServiceUse;
+
+/**
+ * A JUnit 4 Rule to depend on OSGi services.
+ * <p>
+ * Example: <br>
+ *
+ * <pre>
+ * {@literal @}Rule
+ * public ServiceUseRule&lt;Foo> foos = new ServiceUseRule.Builder&lt;>(Foo.class)
+ *     .filter("(service.description=Acme Inc.)")
+ *     .requiredWithin(300)
+ *     .build();
+ * </pre>
+ *
+ * @param <T> the service type
+ */
+public class ServiceUseRule<T> implements AutoCloseable, TestRule {
+
+	public static class Builder<T> {
+
+		private final Class<T>			serviceType;
+		private final BundleContextRule	bundleContextRule;
+		private Filter					filter;
+		private int						cardinality	= 1;
+		private long					timeout		= ServiceUse.DEFAULT_TIMEOUT;
+
+		/**
+		 * @param serviceType of the service
+		 */
+		public Builder(Class<T> serviceType, BundleContextRule bundleContextRule) {
+			this.serviceType = requireNonNull(serviceType);
+			this.bundleContextRule = requireNonNull(bundleContextRule);
+			this.filter = format("(objectClass=%s)", serviceType.getName());
+		}
+
+		/**
+		 * Filter string used to target more specific services.
+		 *
+		 * @param filter string used to target more specific services
+		 */
+		public Builder<T> filter(String filter) {
+			this.filter = format("(&%s%s)", this.filter.toString(), requireNonNull(filter));
+			return this;
+		}
+
+		/**
+		 * Require first matching service to arrive within the default timeout
+		 * {@link ServiceUse#DEFAULT_TIMEOUT}}
+		 */
+		public Builder<T> cardinality(int cardinality) {
+			if (cardinality < 0) {
+				throw new IllegalArgumentException("cardinality must be zero or greater");
+			}
+			this.cardinality = cardinality;
+			return this;
+		}
+
+		/**
+		 * Indicate require services must arrive within the specified timeout.
+		 *
+		 * @param timeout the timeout after which {@link AssertionError} is
+		 *            thrown
+		 */
+		public Builder<T> timeout(long timeout) {
+			if (timeout < 1) {
+				throw new IllegalArgumentException("timeout must be zero or greater");
+			}
+			this.timeout = timeout;
+			return this;
+		}
+
+		public ServiceUseRule<T> build() {
+			return new ServiceUseRule<>(serviceType, bundleContextRule, filter, cardinality, timeout);
+		}
+
+	}
+
+	private final ServiceUse<T> use;
+	private final Class<T>			serviceType;
+	private final BundleContextRule	bundleContextRule;
+
+	protected ServiceUseRule(Class<T> serviceType, BundleContextRule bundleContextRule, Filter filter, int cardinality,
+		long timeout) {
+		this.serviceType = serviceType;
+		this.bundleContextRule = bundleContextRule;
+		this.use = new ServiceUse<>(filter, cardinality, timeout);
+	}
+
+	public void init(Class<?> testClass) {
+		bundleContextRule.init(testClass);
+		use.init(bundleContextRule.getBundleContext());
```

Comment:
> Also, isn't the call to bundleContextRule.init() redundant? The BundleContextRule's own evaluate() statement will call init() with the appropriate class.

Sadly, no. This is the issue with rules in junit4, unlike junit5 extensions, their order of execution (and not instantiation) is unpredictable. Therefore the only way to ensure that init was actually called is to call it. But I've made sure it's idempotent to call it multiple times to not wast energy.

Reply:
You can chain the rules together and guarantee execution order that way by having `ServiceUseRule.apply()` call `BundleContextRule.apply()` and wrapping the resulting statement.:

```
public Statement apply(Statement inner, Description desc) {
    return bcrRule.apply(new Statement() {
      @Override
      public void evaluate() throws Throwable {
        init(description.getTestClass());
        try {
          inner.evaluate();
	} finally {
	  close();
	}
      }
    }, desc;
}

Now it will be guaranteed that bcrRule.apply() is called first.

This is effectively what the `@RuleChain` rule does. Users then need to know not to annotate `BundleContextRule` with the `@Rule` annotation.